### ISSUE TITLE
systrytoopenone deprecated?? (rev. 9f8834f2)

### ISSUE THREAD

**Issue by agraef** [pure-data/pure-data#2416]
Hi @millerpuckette, this is Albert from the JGU in Germany. I'm referring to your commit https://github.com/pure-data/pure-data/commit/9f8834f2083aafd505c6b5ebfc7b1e6ec5e98657.

Can you please explain what this is about? Because it seems that you just renamed the function. Is systrytoopenone / systrytoopenit going away? Or becoming unavailable to externals and loaders? If so, why?

I'm just trying to understand what's happening, and how I'd need to adjust the source of pd-lua so that we don't run into trouble with the next Pd release.

If there's a reason why pd-lua shouldn't use this call anymore then please let me know. Then I just need to write my own replacement for this function.

Quoting from the commit message:

> This can go away once all old copies of pdlua in the world have been updated.

I'm not sure what you mean. By definition, those old pd-lua versions aren't maintained any more.

I've been maintaining pd-lua since 0.8.0, we're at 0.12.8 now. This project is very much alive and kicking. There have been many important contributions by various people in the Pd community. There's a Deken package which now gets updated with every release, and the latest pd-lua version is also included in plugdata and purr-data.

I need to accommodate all of these, so I'd much appreciate it if such breaking API changes were avoided. Unless there's a good reason for them, of course.

**Comment by Lucarda** [comment]
hi @agraef 

see [this](https://github.com/pure-data/pure-data/commit/95f63eb29fccd81ec188c1a0f2701a7898270939#commitcomment-146038615). it has answers to your questions.

**Comment by millerpuckette** [comment]
Hi Graef -

I'm delighted that you're keeping pd-lua well up to date.  My idea of 
the moment is to keep the old function as it was in 0.55-0 so that 
existing versios of pd-lua can cintinue to work.  At your leisure you 
can switch to the newer one.  The old one might go away in, oh, 2030 or 
so :)

Miller

**Comment by agraef** [comment]
Thanks @Lucarda, that indeed explains it. :) I missed that bit with systrytoopenone() growing a new argument.

@millerpuckette:
> At your leisure you can switch to the newer one.

Hi Miller, thanks for taking the time to reply! I can take care of the API change in pd-lua and pd-faustgen right away. But can you please also bump the PD_BUGFIX_VERSION right now, or define some other preprocessor symbol, so that it's possible, you know, to actually _detect_ the API change without using autoconf? Thanks.

> The old one might go away in, oh, 2030 or so :)

Sounds good! We'll both be retired by then, I guess. :)

**Comment by agraef** [comment]
Ok, I have a fix now in both pd-lua and pd-faustgen which checks for version 0.55.1+, but I can't really test it yet until 0.55.1 comes out.

**Comment by millerpuckette** [comment]
Yep... time to call it 0.55-1test1.  I think everything's ready now.  
Many thanks to Iohannes and Christof for preventing me from putting out 
a bogus bugfix a day or two ago.

cheers

M

**Comment by agraef** [comment]
Thanks. Just tested my fix against 0.55.1test1, works.

But there's a big snag with switching out-of-tree externals to the new API function if 0.55.1+ is detected. Those binaries will not load with 0.55.0 and older because they don't have the sys_trytoopenit function. On the other hand, if I use sys_trytoopenone with 0.55.1, I'm getting those obnoxious error messages that you planted there -- and lots of them, that darn message is repeated each and every time sys_trytoopenone is called.

With the Deken packages, we absolutely need to maintain backward compatibility with older Pd versions. So switching to the new API function is not an option right now, and for some time to come.

The only way to work around that in a 3rd party external is to either:
* Use dlsym() to find out what API function to use, which adds a lot of complication.
* Replicate the code for sys_trytoopenone in the external to get rid of the error message, which is a maintenance nightmare.

I want to do neither, they're both horrible kludges. So please *please* remove that error message. I mean those two lines here:
https://github.com/pure-data/pure-data/blob/0146956031eed38d514baf37a1fcc24d4beb85c4/src/s_path.c#L327-L328

At least for now. You can think about getting it back in place to force external developers to actually start using the new API function when Pd 0.55.0 is long dead and forgotten. But as long as these versions are still around on many systems, the old sys_trytoopenone needs to stay. *Without* the error message.

**Comment by millerpuckette** [comment]
Well, OK, I think I can just have a warning message only appear the 
first time sys_trytoopenone() is called.. and then only if we're running 
0.56 or higher :)

**Comment by millerpuckette** [comment]
OK... I've already compiles 0.55-1test1 so this won't make it out to the 
executable, but it's in git repo and should appear for the "real" 0.55-1.

Meanwhile I can't upload 0.55-1test1 til later today for firewall 
reasons... stay tuned :)

cheers

M

**Comment by agraef** [comment]
> and then only if we're running 0.56 or higher :)

I think that you're overly optimistic there. ;-) 0.56 will presumably come out some time next summer? So we get a one-year grace period. Better, but not great.

Miller, I'm not trying to be obnoxious or anything, I just firmly believe that this error message is a horrible mistake. And I will tell you exactly why, so please bear with me while I try to argue my point.

As a 3rd party package maintainer, I'm not worried about the rolling Linux distro, Mac, or Windows users of Pd at all, they can update to the latest Pd at any time, no problem. I'm worried about the Debian stable and Ubuntu LTS (and derivatives) crowd. Case in point: Even Ubuntu 24.04 LTS only ships Pd 0.54.1, and that we will have to support with Deken packages for quite some time, at the *very* least until 26.04 LTS comes out. So we'll need to build Deken packages that use the pre-0.55.1 API at least until then.

Thus a one year grace period until the obnoxious errors come at us again (even though you thankfully fixed them so that at least they don't come in droves anymore), is just way too short.

Why not just let the old API function (without the error message) stick around until you decide it's time to remove it? At which point those old externals will both become unloadable in the latest Pd, and get compile errors when building against the new API, so users of those externals will file bug reports and the developers will come around and fix them. Simple as that.

Also, I don't see why just keeping the old API function around forever would be such a terrible idea. It's not a substantial amount of technical debt, just a one-liner. I'm sure there are parts of the API that are way harder to maintain. :)

Please don't try to stick it to us poor 3rd party external developers with those silly error messages. What lesson are you trying to teach us with those? Presumably most people affected by this API change are expert developers, we *know* how to fix our stuff when it breaks.

Those errors explicitly talk about sys_trytoopenone, a C function deep down in Pd's C API which most users, even developers will probably never have heard about. Those error messages will just scare our users, and I can already see that barrage of bug reports coming because people think that pd-lua or pd-faustgen is broken in some way.

What you *could* do, however, at least with gcc, is to add `__attribute__ ((deprecated))` to the sys_trytoopenone() declaration in the headers. I'm sure you know about this, see [deprecated function attribute](https://gcc.gnu.org/onlinedocs/gcc-14.2.0/gcc/Common-Function-Attributes.html#index-deprecated-function-attribute) in the gcc docs. 

*That* is the proper way to flag an API function as deprecated so that external developers (at least those of us who are worth their salt) will take notice.

TL;DR: Please get rid of that error message, and maybe add a `__attribute__ ((deprecated))` warning instead. The 3rd party external developers will thank you.

I can send you a PR for that, just let me know.

**Comment by Spacechild1** [comment]
> TL;DR: Please get rid of that error message, and maybe add a __attribute__ ((deprecated)) warning instead. The 3rd party external developers will thank you.

Pd has recently added the `PD_DEPRECATED` macro for this purpose.

https://github.com/pure-data/pure-data/blob/cf964806075d9457124f6756367ce575c9a4a800/src/m_pd.h#L93C9-L102

We should really use this macro and not just silently deprecate API functions.

---

That being said, `int sys_trytoopenone` is part of `s_stuff.h` which is *not* part of the public API. The function isn't even marked as `EXTERN` and it is just due to an oversight in Pd's build system that the function is being exported in the first place (see https://github.com/pure-data/pure-data/issues/569). In fact, if you build Pd with MSVC, `sys_trytotopenone` wouldn't be exported.

@agraef I would rather recommend that you find your own replacement for `sys_trytoopenone` and put out a bugfix release. It is not the first time that (private) functions in `s_stuff.h` have changed in the past. For example, many of the audio API functions changed during the big audio settings refactoring and as a consequence the `mediasettings` library had to adept these changes. In the future, please don't use any non-`EXTERN` functions in your externals - or live with the consequences :) (We should actually put a big disclaimer in the headers...)

**Comment by agraef** [comment]
> That being said, int sys_trytoopenone is part of s_stuff.h which is not part of the public API.

I'm aware of that. But it's being used by 3rd party loader externals since those things exist. If you have a function around for that long, it effectively becomes part of the public API. And it's not just some random API function, it's pretty much an essential corner piece of loaders that need to find their source files on Pd's path.

> In fact, if you built Pd with MSVC, sys_trytotopenone wouldn't be exported!

But the official packages by Miller obviously aren't, and that's what we always build against on Windows. So the point is moot.

> I would rather recommend that you find your own replacement for sys_trytoopenone and put out a bugfix release.

I will, if I'm forced to. But it's a very bad idea. We really want 3rd party loaders to use the same search algorithm that Pd's own loader uses. Otherwise each and every loader will have to ship with their own sys_trytoopenone implementation, which will breed inconsistencies.

**Comment by agraef** [comment]
> If the function is that important then we should stabilize it and move it to m_pd.h!

Exactly! :)

If I'm honest, I really don't understand why we're arguing about this at all. sys_trytoopenone is essentially just a one-liner now. Please keep it around is all I'm saying. Everything else just makes the lives of developers of 3rd party loader  externals (which IMHO are an important advantage of Pd over other similar environments) miserable for no good reason.

**Comment by Spacechild1** [comment]
> If I'm honest, I really don't understand why we're arguing about this at all. 

Because we should really draw a clear distinction between the public and private API. We *need* to be able to freely break the private API and not worry that we might accidently break some external. https://github.com/pure-data/pure-data/issues/569 would really help with that because then externals simply can't link against private functions anymore.

> Please keep it around is all I'm saying. 

I'm ok with keeping it around, until we have a proper public API function.

**Comment by sebshader** [comment]
sure, loader developers could just include their own version of sys_trytoopenone. But wouldn't that just prevent any issues that are found with the function from being fixed in externals automatically? external authors can't be sure about the threading/logging context these functions will be called in so imo it makes some sense for there to be public api for it based on that usage/consideration.

**Comment by Spacechild1** [comment]
I guess I was only a bit angry because you accused us of an API break and demanded a fix although the function was never part of the public API to begin with...

Anyway, I just saw that Miller has reduced the verbosity of `sys_trytopenone`. Note, however, that you still have to update the Pd sources because the function declaration of `sys_trytoopenone` has been removed from `s_stuff.h`.

@millerpuckette Can we take the chance and make `sys_trytoopenit` (= the new function) public? (Maybe with a better name? :)
The function itself is not particularly interesting - there is no "search algorithm" -, but it does some environment variable expansion and posts some verbose messages. Therefore I agree with @agraef that it's useful for making custom loaders behave like the built-in loader.

**Comment by agraef** [comment]
> I guess I was only a bit angry because you accused us of an API break and demanded a fix although the function was never part of the public API to begin with...

That was never my intention, sorry. I *am* aware that it's not part of the official API (yet), and of course I understand the importance of distinguishing between private and public APIs.

I guess in the end I was just arguing that we absolutely need this as a public API function for the external loader developers at least.

> I'm ok with keeping it around, until we have a proper public API function.

Ok, I understand that you need to discuss this. So for the time being I will assume that sys_trytoopenone is there to stay, and that the error message that's now in there will eventually be removed.

> Note, however, that you still have to update the Pd sources because the function declaration of sys_trytoopenone has been removed from s_stuff.h.

Ok, I can just conditionally add an extern declaration for that function in pdlua.c, no problem. Same for pd-faustgen.

> (Maybe with a better name? :)

Please don't do that. We need the function to stay exactly as it is, *including its name*. Otherwise we either break backwards or forwards compatibility, as I explained above. Hence, changing the name would be just as bad as removing the function entirely.

**Comment by Spacechild1** [comment]
> Ok, I can just conditionally add an extern declaration for that function in pdlua.c, no problem. Same for pd-faustgen.

+1

> Please don't do that. We need the function to stay exactly as it is, including its name. Otherwise we either break backwards or forwards compatibility, as I explained above. Hence, changing the name would be just as bad as removing the function entirely.

I'm talking about the *new* function (= `sys_trytoopenit`). This one has added just a few days ago, so I don't think that any externals already rely on it :)

**Comment by agraef** [comment]
> I'm talking about the new function (= sys_trytoopenit).

Haha, that was a good one. You really had me there.

> This one has added just a few days ago, so I don't think that any externals already rely on it :)

You never know. Don't Pd external developers keep monitoring the header files in git at day and night to find new non-API functions they might (ab)use? :))

**Comment by Spacechild1** [comment]
> Don't Pd external developers keep monitoring the header files in git at day and night to find new non-API functions they might (ab)use? :))

I wouldn't be surprised :D

**Comment by Spacechild1** [comment]
BTW, I made sure that `sys_trytoopenone` is really exported: https://github.com/pure-data/pure-data/pull/2418/commits/963c37641327bd9a0b557559bb448bc6459651ad

**Comment by agraef** [comment]
Cool. So I can now just revert the adjustments I committed in pd-lua and pd-faustgen and pretend that all is fine and dandy again.

**Comment by Spacechild1** [comment]
Just to be clear: you still need the `EXTERN` forward declaration, though. The function is *exported*, but it's not being declared in `s_stuff.h` anymore.

**Comment by agraef** [comment]
Ok, got it, thanks.

**Comment by agraef** [comment]
Alright, I think that I got this all sorted on my side for now. @Spacechild1 thanks a lot for joining the discussion and helping me better understand the issue and what my options are if sys_trytoopenone keeps being marked as obsolete. I guess I just keep this ticket open for now, until the future status of that function has been decided.
